internal/cas/content.go

```
@@ -0,0 +1,224 @@
+
+		if readErr != nil {
+			return &WrappedError{
+				Op:   "read_source",
+				Path: sourcePath,
+				Err:  ErrReadFile,
+			}
+		}
+
+		// Write to temporary file first
+		tempPath := targetPath + ".tmp"
+
+		if err := os.WriteFile(tempPath, data, RegularFilePerms); err != nil {
+			return &WrappedError{
+				Op:   "write_target",
+				Path: tempPath,
+				Err:  err,
+			}
+		}
+
+		// Rename to final path
+		if err := os.Rename(tempPath, targetPath); err != nil {
+			return &WrappedError{
+				Op:   "rename_target",
+				Path: tempPath,
+				Err:  err,
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Content) ensureTargetDirectory(targetPath string) error {
+	targetDir := filepath.Dir(targetPath)
+	if err := os.MkdirAll(targetDir, DefaultDirPerms); err != nil {
+		return &WrappedError{
+			Op:   "create_target_dir",
+			Path: targetDir,
+			Err:  ErrCreateDir,
+		}
+	}
+
+	return nil
+}
+
+// Store stores a single content item
+func (c *Content) Store(l *log.Logger, hash string, data []byte) error {
+	c.store.mapLock.Lock()
+
+	if _, ok := c.store.locks[hash]; !ok {
+		c.store.locks[hash] = &sync.Mutex{}
+	}
+
+	c.store.locks[hash].Lock()
+	defer c.store.locks[hash].Unlock()
+
+	c.store.mapLock.Unlock()
+
+	if c.store.HasContent(hash) {
+		return nil
+	}
+
+	if err := os.MkdirAll(c.store.Path(), DefaultDirPerms); err != nil {
+		return wrapError("create_store_dir", c.store.Path(), ErrCreateDir)
+	}
+
+	// Ensure partition directory exists
+	partitionDir := c.getPartition(hash)
+	if err := os.MkdirAll(partitionDir, DefaultDirPerms); err != nil {
+		return wrapError("create_partition_dir", partitionDir, ErrCreateDir)
+	}
+
+	path := c.getPath(hash)
+	tempPath := path + ".tmp"
+
+	f, err := os.OpenFile(tempPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, StoredFilePerms)
+	if err != nil {
+		return wrapError("create_temp_file", tempPath, err)
+	}
+
+	buf := bufio.NewWriter(f)
+
+	if _, err := buf.Write(data); err != nil {
+		f.Close()
```

Looks like the returned error is ignored